winnowing.py

```diff
@@ -3,122 +3,134 @@
+                    newStart = self.kGrams1[match][2]   #start position of matched k-gram in cleaned up code
+                    newEnd = self.kGrams1[match][3]   #end position
+                    for k in self.token1:
+                        if k[2] == newStart:   #linking positions in cleaned up code to original code
+                            startx = k[1]
+                            flag = 1
+                        if k[2] == newEnd:
+                            endx = k[1]
+                    if flag == 1:
+                        points.append([startx, endx])
+        points.sort(key = lambda x: x[0])
+        points = points[1:]
+        mergedPoints = []
+        mergedPoints.append(points[0])
+        for i in points:
+            last = mergedPoints[-1]
+            if i[0] >= last[0] and i[0] <= last[1]: #merging overlapping regions
+                if i[1] > last[1]:
+                    mergedPoints = mergedPoints[:-1]
+                    mergedPoints.append([last[0], i[1]])
+                else:
+                    pass
+            else:
+                mergedPoints.append(i)
+        newCode = code[: mergedPoints[0][0]]
+        plagCount = 0
+        for i,value in enumerate(mergedPoints):
+            if value[1] > value[0]:
+                plagCount += value[1] - value[0]
+                newCode = newCode + '\x1b[6;30;42m' + code[value[0] : value[1]] + '\x1b[0m'
+                if i < len(mergedPoints) - 1:
+                    newCode = newCode + code[value[1] : mergedPoints[i+1][0]]
+                else:
+                    newCode = newCode + code[value[1] :]
+        return {"ratio":(plagCount/len(code)),"Code":newCode}
 
-#we form windows of hash values and use min-hash to limit the number of fingerprints
-def fingerprints(arr, winSize = 4):
-    arrLen = len(arr)
-    prevMin = 0
-    currMin = 0
-    windows = []
-    fingerprintList = []
-    for i in range(arrLen - winSize):
-        win = arr[i: i + winSize]  #forming windows
-        windows.append(win)
-        currMin = i + minIndex(win)
-        if not currMin == prevMin:  #min value of window is stored only if it is not the same as min value of prev window
-            fingerprintList.append(arr[currMin])  #reduces the number of fingerprints while maintaining guarantee
-            prevMin = currMin  #refer to density of winnowing and guarantee threshold (Stanford paper)
 
-    return fingerprintList
+    def jaccardCheck(self):
+        fp1 = set(self.fpList1)
+        fp2 = set(self.fpList2)
+        comman = fp1.intersection(fp2)
+        total = len(fp1)+len(fp2)
+        plagcount = len(comman)
+        return {"ratio":(plagcount/(total-plagcount))}
+        
+            
+    def plagiarismRate(self):
+        total = len(self.fpList1)
+        plagcount =0
+        for i in self.fpList1:
+            if i in self.fpList2:
+                plagcount+=1
```

Comment:
Instead of list for make them sets and use intersection. And get the length of matched elements from intersection. Apply this wherever possible

Reply:
done sir